### PR TITLE
candidate fix for v18

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Needs to be cherry-picked onto `main`

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: app/upgrades/v18/upgrades_test.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	var lastPoolID uint64 // To keep track of the last assigned pool ID
+
+	// Sort AssetPairs based on LinkedClassicPool values.
+	// We sort both pairs because we use the test asset pairs to create initial state,
+	// then use the actual asset pairs to verify the result is correct.
+	sort.Sort(ByLinkedClassicPool(v17.AssetPairsForTestsOnly))
+	sort.Sort(ByLinkedClassicPool(v17.AssetPairs))
+
+	// Create earlier pools or dummy pools if needed
+	for _, assetPair := range v17.AssetPairsForTestsOnly {
```
Comment (p0mvn) [inline app/upgrades/v18/upgrades_test.go:90]: Why do we not iterate over `v17.AssetPairs`, is that intended?

File: app/upgrades/v18/upgrades_test.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	suite.Ctx = suite.Ctx.WithBlockTime(suite.Ctx.BlockTime().Add(time.Hour * 24))
+
+	// after the accum values have been resetted correctly after upgrade, we expect the accumulator store to be initialized with the correct value,
+	// which in our test case would be 10000(the amount that was locked)
+	valueAfterClear := suite.App.LockupKeeper.GetPeriodLocksAccumulation(suite.Ctx, lockuptypes.QueryCondition{
+		LockQueryType: lockuptypes.ByDuration,
+		Denom:         "gamm/pool/3",
+		Duration:      time.Hour * 24 * 14,
+	})
+	valueAfterClear.Equal(sdk.NewInt(shareStaysLocked))
```
Comment (p0mvn) [inline app/upgrades/v18/upgrades_test.go:65]: Can we also run an epoch end hook and see that it does not panic?

File: app/upgrades/v18/upgrades_test.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+func assertEqual(suite *UpgradeTestSuite, pre, post interface{}) {
+	suite.Require().Equal(pre, post)
+}
+
+func (suite *UpgradeTestSuite) TestUpgrade() {
+	// set up pools first to match v17 state(including linked cl pools)
+	suite.setupPoolsToMainnetState()
+
+	// corrupt state to match mainnet state
+	suite.setupCorruptedState()
```
Comment (p0mvn) [inline app/upgrades/v18/upgrades_test.go:51]: Can we also run an epoch end hook and seeing that it does panic?

File: x/incentives/keeper/distribute.go [hunk 4]
```diff
@@ -358,6 +358,22 @@ func (k Keeper) distributeInternal(
...
+			return totalDistrCoins, err
+		}
+
+		// Remove some spam gauges, is state compatible.
+		// If they're to pool 1 they can't distr at this small of a quantity.
+		if remainCoins.Len() == 1 && remainCoins[0].Amount.LTE(sdk.NewInt(10)) && gauge.DistributeTo.Denom == "gamm/pool/1" && remainCoins[0].Denom != "uosmo" {
+			ctx.Logger().Debug(fmt.Sprintf("gauge debug, this gauge is perceived spam, skipping %d", gauge.Id))
+			err := k.updateGaugePostDistribute(ctx, gauge, totalDistrCoins)
+			return totalDistrCoins, err
+		}
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:375]: This is the state-compatible fix to speed up incentive distribution. Makes sense to me to include it

File: x/lockup/keeper/lock.go [hunk 5]
```diff
@@ -295,6 +295,22 @@ func (k Keeper) clearKeysByPrefix(ctx sdk.Context, prefix []byte) {
 	}
 }
 
+func (k Keeper) RebuildAccumulationStoreForDenom(ctx sdk.Context, denom string) {
```
Comment (p0mvn) [inline x/lockup/keeper/lock.go:298]: TODO: unit test this if time

File: app/upgrades/v18/upgrades_test.go (hunk 2 above)
Comment (mattverse) [reply app/upgrades/v18/upgrades_test.go:72]: Added in https://github.com/osmosis-labs/osmosis/pull/6144/commits/9815fa0bf1171e23eededa08b16b749c400d5e8f

File: app/upgrades/v18/upgrades_test.go (hunk 3 above)
Comment (mattverse) [reply app/upgrades/v18/upgrades_test.go:52]: Added in https://github.com/osmosis-labs/osmosis/pull/6144/commits/9815fa0bf1171e23eededa08b16b749c400d5e8f

File: app/upgrades/v18/upgrades_test.go (hunk 1 above)
Comment (mattverse) [reply app/upgrades/v18/upgrades_test.go]: Fixed in https://github.com/osmosis-labs/osmosis/pull/6144/commits/9815fa0bf1171e23eededa08b16b749c400d5e8f

File: x/incentives/keeper/distribute.go [hunk 6]
```diff
@@ -330,7 +330,7 @@ func (k Keeper) distributeInternal(
 			// Note: reason why we do millisecond conversion is because floats are non-deterministic.
 			emissionRate := sdk.NewDecFromInt(remainAmountPerEpoch).QuoTruncate(sdk.NewDec(currentEpoch.Duration.Milliseconds()).QuoInt(sdk.NewInt(1000)))
 
-			ctx.Logger().Debug("distributeInternal, CreateIncentiveRecord NoLock gauge", "module", types.ModuleName, "gaugeId", gauge.Id, "poolId", pool.GetId(), "remainCoinPerEpoch", remainCoinPerEpoch, "height", ctx.BlockHeight())
+			ctx.Logger().Info("distributeInternal, CreateIncentiveRecord NoLock gauge", "module", types.ModuleName, "gaugeId", gauge.Id, "poolId", pool.GetId(), "remainCoinPerEpoch", remainCoinPerEpoch, "height", ctx.BlockHeight())
```
Comment (ValarDragon) [inline x/incentives/keeper/distribute.go:333]:
```suggestion
			ctx.Logger().Debug("distributeInternal, CreateIncentiveRecord NoLock gauge", "module", types.ModuleName, "gaugeId", gauge.Id, "poolId", pool.GetId(), "remainCoinPerEpoch", remainCoinPerEpoch, "height", ctx.BlockHeight())
```

**Review by nicolaslara — APPROVED** [review]

**Review by ValarDragon — APPROVED** [review]
Tested this on a flow where we:
Make a v18 with broken accum store
Make a v19 with these changes

We saw that v18 caused the break as expected, and v19 fixed it. 

Hence this as v19 should fix the mainnet problem.

The load bearing part of the fix is actually ran every testnet as InitGenesis

**Pull request merged**